File: answer/82/model_answer.go
/*
 * @lc app=leetcode id=82 lang=golang
 *
 * [82] Remove Duplicates from Sorted List II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func deleteDuplicates(head *ListNode) *ListNode {
	dummy := &ListNode{0, head}
	previous := dummy
	current := head
	for current != nil {
		nextNode := current.Next
		for nextNode != nil && current.Val == nextNode.Val {
			nextNode = nextNode.Next
		}
		if current.Next == nextNode {
			// 値が変わっていないということ
			previous = current
		} else {
			previous.Next = current
		}
		current = nextNode
	}
	return dummy.Next
}
// @lc code=end

File: answer/2/model_answer.cpp
/*
 * @lc app=leetcode id=2 lang=cpp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
//  struct ListNode {
//       int val;
//       ListNode *next;
//       ListNode() : val(0), next(nullptr) {}
//       ListNode(int x) : val(x), next(nullptr) {}
//       ListNode(int x, ListNode *next) : val(x), next(next) {}
//   };

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode *node = &dummy;
        int val = 0;
        while (l1 || l2 || val)
        {
            val += (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
            node->next = new ListNode(val % 10);
            val /= 10;
            node = node->next;
            if (l1) l1 = l1->next;
            if (l2) l2 = l2->next;
        }
        return dummy.next;
    }
// @lc code=end

